FLT/Deformations/Algebra/InverseLimit/Basic.lean

```diff
@@ -0,0 +1,472 @@
+/-
+Copyright (c) 2025 Javier López-Contreras. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Javier López-Contreras
+-/
+import Mathlib.Algebra.Module.LinearMap.Defs
+import Mathlib.Data.Rat.Cast.Defs
+import Mathlib.Order.DirectedInverseSystem
+import Mathlib.Tactic.SuppressCompilation
+
+/-!
+# Inverse limit of algebraic structures
+
+We introduce all kinds of algebraic instances on `InverseLimit`, and specialize to the cases
+of modules and rings, showing that they are indeed limits in the respective categories.
+-/
+
+suppress_compilation
+
+variable {R ι : Type*} [Preorder ι] {G : ι → Type*}
+variable {T : ∀ ⦃i j : ι⦄, i ≤ j → Type*} {f : ∀ _ _ h, T h}
+variable [∀ i j (h : i ≤ j), FunLike (T h) (G j) (G i)] [InverseSystem (f · · ·)]
+variable [IsDirected ι (· ≤ ·)]
+
+variable (G f) in
+def InverseLimit : Type _ := {x : (i : ι) → G i // ∀ i j h, f i j h (x j) = (x i)}
+
+namespace InverseLimit
+
+variable (i : ι) (x y z : InverseLimit G f)
+
+instance : CoeOut (InverseLimit G f) ((i : ι) → G i) where
+  coe x := x.val
+
+omit [InverseSystem fun x1 x2 x3 ↦ ⇑(f x1 x2 x3)] [IsDirected ι fun x1 x2 ↦ x1 ≤ x2] in
+@[simp]
+lemma prop : ∀ (i j : _) (h : i ≤ j), f i j h (x.1 j) = x.1 i := x.2
+
+omit [InverseSystem fun x1 x2 x3 ↦ ⇑(f x1 x2 x3)] [IsDirected ι fun x1 x2 ↦ x1 ≤ x2] in
+@[ext]
+lemma ext_lemma (h : ∀ i, x.1 i = y.1 i) : x = y := by
+  unfold InverseLimit at *
```

Comment:
This is kind of a code smell. Why not make `InverseLimit_def`, prove it with `rfl`, and then `rw [InverseLimit_def]`?